ext/nnstreamer/tensor_filter/tensor_filter_movidius_ncsdk2.c

```diff
@@ -24,44 +24,248 @@
  * @bug     No known bugs except for NYI items
  *
  * This is the per-NN-framework plugin (Intel Movidius NCSDK2) for tensor_filter.
- * TODO: Fill in "GstTensorFilterFramework" for tensor_filter.h/c
- *
  */
 
+#include <fcntl.h>
 #include <glib.h>
+#include <glib/gstdio.h>
+#include <gst/gst.h>
 #include <mvnc2/mvnc.h>
 #include <nnstreamer_plugin_api_filter.h>
+#include <sys/types.h>
+#include <sys/stat.h>
 
 void init_filter_mvncsdk2 (void) __attribute__ ((constructor));
 void fini_filter_mvncsdk2 (void) __attribute__ ((destructor));
 
+enum _private_constants
+{
+  NNS_MVNCSDK2_SUPPORT_MAX_NUMS_DEVICES = 8,
+  NNS_MVNCSDK2_SUPPORT_API_MAJOR_VER = 2,
+  NNS_MVNCSDK2_API_VER_ARRAY_SIZE = NC_VERSION_MAX_SIZE,
+  NNS_MVNCSDK2_MAX_NUM_ELEM_IN_FIFO = 2,
+};
+
+static const char NNS_MVNCSDK2_NAME_INPUT_FIFO[] = "INPUT_FIFO";
+static const char NNS_MVNCSDK2_NAME_OUTPUT_FIFO[] = "OUTPUT_FIFO";
+
 /**
  * @brief internal data of mvncsdk2
  */
 typedef struct _mvncsdk2_data
 {
-  void *private_data;
+  /* Variables of the data types from mvnc.h */
+  struct ncDeviceHandle_t *handle_device; /** handle for Intel Movidius device */
+  struct ncGraphHandle_t *handle_graph; /** handle for graph (model) */
+  struct ncTensorDescriptor_t tensor_desc_input; /** description of input tensor */
+  struct ncTensorDescriptor_t tensor_desc_output; /** description of output tenser */
+  struct ncFifoHandle_t *handle_fifo_input; /** handle for input fifo (buffer) */
+  struct ncFifoHandle_t *handle_fifo_output; /** handle for output fifo (buffer) */
+  /* Normal variables */
+  GMappedFile *gmapped_file_model; /** a mapped file for model */
+  gint32 idx_device;  /** index of device to use (Q. is it necessary?) */
 } mvncsdk2_data;
 
 /**
  * @brief Free privateData and move on.
+ * @param prop : property of tensor_filter instance
+ * @param private_data : movidius-ncsdk2 plugin's private data
  */
 static void
 _mvncsdk2_close (const GstTensorFilterProperties * prop, void **private_data)
 {
-  return;
+  mvncsdk2_data *pdata = *private_data;
+
+  if (pdata != NULL) {
+    ncFifoDestroy (&(pdata->handle_fifo_input));
+    ncFifoDestroy (&(pdata->handle_fifo_output));
+    ncGraphDestroy (&(pdata->handle_graph));
+    ncDeviceDestroy (&(pdata->handle_device));
+    g_mapped_file_unref (pdata->gmapped_file_model);
```

Comment:
I think the refcount of  `pdata->gmapped_file_model` is not increased before.

Reply:
@dongju-chae IMHO, g_mapped_file_new( ) increases the ref count.

Reply:
Yeap, in old version, gmapped_file_model was null (not assigned). It seems it's not used now :)